docs/api-proposal-2.md

```diff
@@ -0,0 +1,113 @@
+```
+VPC: Vpc-1 (existing definition)
+  PifX:
+    Internal: [10.0.1.0/24] // these IPs may communicate through this pif
+    External: [192.168.1.0/24] //... using addresses in this range
+    Mapping function:  
+		10.0.1.0/24 -> 192.168.1.0/24
+
+    explanation: let any IP in 10.0.1.0/24 use this pif to reach to the destinations 
+                 at any the other end(s) of the virtual link(s) it is connected to 
+                 by mapping the addresses to any address in 192.1681.0/24.
+
+    // other stuff that could go in a pif
+    Max-rate: 200MB/s
+    Qos: some-user-defined-qos-object (optional)
+    Filtering:
+       Ingress: user-filtering-rule-object (optional)
+       Egress:  user-filtering-rule-object (optional)
+
+VPC: Vpc-2 (existing definition)
+  PifY:
+    Internal: 10.0.2.0/24
+    External: 192.168.2.0/24
+    Mapping function: 
+       1) 10.0.2.1 -> 192.168.2.29
+       2) 10.0.2.2 <-> 192.168.2.30
+       3) 10.0.2.8/30 <- 192.168.2.31
+       4) 10.0.2.128/25 -> 192.168.2.128/25
+
+	explanation:
+      1) 10.0.2.1 can go outside as 192.168.2.29 (snat)
+      2) 10.0.2.2 can go outside as 192.168.2.30(snat). 
+         It can be contacted externally at 192.168.2.30
+      3) 10.0.2.[8,9,10,11] can be contacted from outside as 192.168.2.31
+      4) any ip in 10.0.2.128/25 will use any ip from 192.168.2.128/25 as external identity.
+      5) 10.0.2.0/25 except for 10.0.2.[1,2,8,9,10,11] will not be reachable from outside nor are 
+         allowed to get outside since the mapping function did not cover them
+
+// at a distinct scope (Fabric)
+Virtual links:
+    left: 
+      vpc: Vpc-1
+      interface: PifX
+    Right: 
+      vpc: VPC-2
+      interface: PifY
+```
+
+
+
+## How it differs from what was specified in the past
+* no communication semantics or patterns. These are in the mapping function. Virtual links just define potential reachability.
+* External addresses, while referred to by PIFs are not exclusive of  particular PIFs but VPCs. This is an important difference affecting resource usage.
+
+## Pros of this model (IMO):
+* feels natural to users: they plug cables and specify what can flow through them. Pifs indicate egress / ingress points.
+* Having an interface object also feels natural, demarcates ownership (needed for multi-tenancy) and acts as an anchor point to extend the model in the future or provide certain functions (rate limitting, but also traffic inspection for instance).
+* Can't be too distinct from the implementation. The mapping function may seem cumbersome, but may resemble what the gateway will do after all. It gives significant flexibility.
+* The model is easy. The complexity (if any) lies in the mapping function. That's a single place where users shall concentrate care, which they should/will.
+
+## Constraints
+* An external addresses can belong to only one VPC.
+* No other restriction is needed IMO, but happy to discuss why.
```

Comment:
Can you discuss why?  Dealing with multiple pifs with the same IP addresses and return routes in the gateway was the whole issue here, especially if there are conflicting policies between the PIFs (e.g., firewall rules, PAT, etc.).  Why is this no longer an issue?

Reply:
Yes, my example for the firewall rules at the pif level was not good.
The intent --still valid-- was to show that having an entity that users can refer to and acts as anchor, eases things. Now, for firewall rules, you'd probably have some other entity outside the pifs for that, (which could refer to the pifs if needed) indicating what traffic is allowed.